### DISCUSSION TITLE
Why doesn't my SVG grid work?

### DISCUSSION THREAD

**Discussion by mscofield0** [fabricjs/fabric.js#7707] [Q&A]
So I want to create an infinite grid for my background in the canvas, the problem is that the following code (mostly taken from [here](https://codereview.stackexchange.com/questions/114702/drawing-a-grid-on-canvas)) doesn't work:

```ts
export class Grid implements Renderable {
    static readonly CELL_SIZE: number = 20;

    render(canvas: fabric.Canvas) {
        const svg = `
            <svg width="100%" height="100%" xmlns="http://www.w3.org/2000/svg">
                <defs>
                    <pattern id="smallGrid" width="${Grid.CELL_SIZE}" height="${Grid.CELL_SIZE}" patternUnits="userSpaceOnUse">
                        <path d="M ${Grid.CELL_SIZE} 0 L 0 0 0 ${Grid.CELL_SIZE}" fill="none" stroke="gray" stroke-width="0.5" />
                    </pattern>
                </defs>

                <rect width="100%" height="100%" fill="url(#smallGrid)" />
            </svg>
        `;
        const blob = new Blob([svg], {type: "image/svg+xml;charset=utf-8"});
        const url = URL.createObjectURL(blob);

        fabric.Image.fromURL(url, (img) => {
            img.hasBorders = img.hasControls = false;
            canvas.add(img);
        });
    }
}
```
and then used in the Svelte component:
```ts
onMount(() => {
  canvas = new fabric.Canvas(canvas_el);
  resizeCanvas();

  new Grid().render(canvas);

  return () => canvas.dispose();
});
```

How do I make this work?

**Comment by ShaMan123** [comment]
It's a strange approach, why go through svg?

**Reply by mscofield0** [reply]
The alternative is spawning an arbitrary amount of lines which isn't efficient and isn't really infinite either.

**Reply by ShaMan123** [reply]
You should look into fabric.Pattern

**Reply by mscofield0** [reply]
Using Pattern allowed me to finally load to svg in, but there is an issue currently... It's all blurry for some reason. I tried fiddling around with the cell size variable, and it's blurry in all instances. It is also almost fully black at the beginning, but then I just click the canvas and it stops being fully black.

Fully black:
![image](https://user-images.githubusercontent.com/69715407/154824126-d939fd74-ab7a-4e89-bdcf-3025dd345e6c.png)

After clicking:
![image](https://user-images.githubusercontent.com/69715407/154824121-1d7e5660-a44d-471f-bb3d-9f5fc460aea9.png)


This is the new code:
```ts
import type { Renderable } from "./renderable";
import { fabric } from "fabric";

export class Grid implements Renderable {
    static readonly CELL_SIZE: number = 5;

    render(canvas: fabric.Canvas) {
        // Also tried with `<svg width="100%" height="100%" ...>`
        const svg = `
            <svg width="${Grid.CELL_SIZE}" height="${Grid.CELL_SIZE}" xmlns="http://www.w3.org/2000/svg">
                <defs>
                    <pattern id="smallGrid" width="${Grid.CELL_SIZE}" height="${Grid.CELL_SIZE}" patternUnits="userSpaceOnUse">
                        <path d="M ${Grid.CELL_SIZE} 0 L 0 0 0 ${Grid.CELL_SIZE}" fill="none" stroke="gray" stroke-width="0.5" />
                    </pattern>
                </defs>

                <rect width="100%" height="100%" fill="url(#smallGrid)" />
            </svg>
        `;
        const blob = new Blob([svg], {type: "image/svg+xml;charset=utf-8"});
        const url = URL.createObjectURL(blob);
        const plane = new fabric.Rect({
            width: 100_000,
            height: 100_000,
            left: 0,
            top: 0,
            hasBorders: false,
            hasControls: false,
            selectable: false,
            
        });

        fabric.util.loadImage(url, (img) => {
            plane.set("fill", new fabric.Pattern({
                source: img,
                repeat: "repeat",
            }))
        });

        canvas.add(plane);
    }
}
```

**Reply by ShaMan123** [reply]
Dump the svg and take a look at the demos 
http://fabricjs.com/patterns
http://fabricjs.com/freedrawing (pattern brush)

**Reply by ShaMan123** [reply]
http://fabricjs.com/fabric-gotchas

**Comment by melchiar** [comment]
@ShaMan123 is right, svg is an odd approach. 

Not sure what you're needing the grid for, but if it's just a guide and you don't need to serialize it or include it in dataURLs, I'd grab the drawGrid function from this answer (https://codereview.stackexchange.com/a/135207), and call it from the `before:render` event. With some slight tweaks the grid can be made infinite and adjust for zoom/pan. You'll also need to remove the background color on your canvas and add your own background draw call since backgrounds are drawn after the `before:render` event is called. This would be the cleanest and most performant solution.

**Comment by asturur** [comment]
SVG is an odd approach because while it could render at any resolution, it will get cached as an image and when you zoom it the stroke will grow and you will alignment issues with the pixels. ( the 1 px line will be often across pixels ).

As @melchiar mentioned you can either look into that code for before render, or you can still use a pattern ( still in before render ) with this characteristics:

you can have as backgroundImage a rect that is filled with a pattern.
You would redraw the pattern source on every pan change or zoom change ( Differently but similarly ).
Let me see if i can sketch an example:
https://jsfiddle.net/we2k10t3/6/

This is by no mean a complete example, for which i don't have time.
But it gives you a track.
You add this grid as background image and in the updateGrid function you run your logic to update the pattern to give the impression the pattern is actually scrolling.
I m sure you can figure out the rest, if you can't complete the example with zoom controls and what you tried and we can move forward from there.
